flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(
+			keySerializer,
+			namespaceSerializer,
+			stateSerializer,
+			spaceAllocator,
+			0,
+			1.0f);
+
+		StateSnapshotTransformer<String> transformer = new StateSnapshotTransformer<String>() {
+			@Nullable
+			@Override
+			public String filterOrTransform(@Nullable String value) {
+				if (value == null) {
+					return null;
+				}
+				int op = value.hashCode() % 3;
+				switch (op) {
+					case 0:
+						return null;
+					case 1:
+						return value + "-transform";
+					default:
+						return value;
+				}
+			}
+		};
+
+		// map to store expected states, namespace -> key -> state
+		Map<Long, Map<Integer, String>> referenceStates = new HashMap<>();
+		int totalStateSize = 0;
+		int totalSizeIncludingLogicalRemovedKey = 0;
+		int totalLogicallyRemovedKey = 0;
+		int totalSpaceNumber = 0;
+
+		// take snapshot 1 which is an empty snapshot
+		Map<Long, Map<Integer, String>> expectedSnapshot1 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot1 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+		assertEquals(1, stateMap.getHighestRequiredSnapshotVersion());
+		assertEquals(1, stateMap.getSnapshotVersions().size());
+		assertEquals(true, stateMap.getSnapshotVersions().contains(1));
+		assertEquals(1, stateMap.getResourceGuard().getLeaseCount());
+		verifySnapshotWithoutTransform(
+			expectedSnapshot1, snapshot1, keySerializer, namespaceSerializer, stateSerializer);
+		verifySnapshotWithTransform(
+			expectedSnapshot1, snapshot1, transformer, keySerializer, namespaceSerializer, stateSerializer);
+
+		snapshot1.release();
+		assertEquals(1, stateMap.getStateMapVersion());
+		assertEquals(0, stateMap.getHighestRequiredSnapshotVersion());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+		assertTrue(stateMap.getSnapshotVersions().isEmpty());
+		assertEquals(0, stateMap.getResourceGuard().getLeaseCount());
+
+		// put some states
+		for (int i = 1; i <= 10; i++) {
+			totalStateSize++;
+			totalSizeIncludingLogicalRemovedKey++;
+			totalSpaceNumber += 2;
+			stateMap.put(i, (long) i, String.valueOf(i));
+			addToReferenceState(referenceStates, i, (long) i, String.valueOf(i));
+		}
+
+		// take snapshot 2
+		Map<Long, Map<Integer, String>> expectedSnapshot2 = snapshotReferenceStates(referenceStates);
+		CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot2 =
+			(CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String>) stateMap.stateSnapshot();
+		assertEquals(2, stateMap.getStateMapVersion());
+		assertEquals(2, stateMap.getHighestRequiredSnapshotVersion());
+		assertEquals(1, stateMap.getSnapshotVersions().size());
+		assertEquals(true, stateMap.getSnapshotVersions().contains(2));
+		assertEquals(1, stateMap.getResourceGuard().getLeaseCount());
+
+		// 1. test put -> put -> remove for (key 1, namespace 1)
```

Comment:
As this comment indicates, I think we should split this test up into several and more targeted tests.